simple_connection_events/simple_connection_events.cpp

```diff
@@ -0,0 +1,214 @@
+
+static NabtoClientFuture* eventFuture_;
+static NabtoClientListener* listener_;
+static NabtoClientConnectionEvent connectionEvent_;
+
+void arm_connection_event_listener(NabtoClientConnection* connection);
+
+const char* connection_event_to_string(NabtoClientConnectionEvent event) {
+    if (event == NABTO_CLIENT_CONNECTION_EVENT_CLOSED) {
+        return "Connection closed";
+    } else if (event == NABTO_CLIENT_CONNECTION_EVENT_CONNECTED) {
+        return "Connection opened";
+    } else if (event == NABTO_CLIENT_CONNECTION_EVENT_CHANNEL_CHANGED) {
+        return "Channel changed";
+    } else {
+        return "Unknown event";
+    }
+}
+
+const char* connection_type_to_string(NabtoClientConnection* connection) {
+    NabtoClientConnectionType type;
+    nabto_client_connection_get_type(connection, &type);
+    return type == NABTO_CLIENT_CONNECTION_TYPE_RELAY ? "Relay" : "Direct";
+}
+
+void print_connection_type(NabtoClientConnection* connection) {
+    std::cout << "Connection type is now: " << connection_type_to_string(connection) << std::endl;
+}
+
+void connection_event_cb(NabtoClientFuture* future, NabtoClientError ec, void* data) {
+    if (ec != NABTO_CLIENT_EC_OK) {
+        std::cerr << "Connection event callback error: " << nabto_client_error_get_message(ec) << std::endl;
+    } else {
+        std::cout << "Connection event: " << connection_event_to_string(connectionEvent_) << std::endl;
+        if (connectionEvent_ == NABTO_CLIENT_CONNECTION_EVENT_CHANNEL_CHANGED) {
+            print_connection_type((NabtoClientConnection*)data);
+        }
+    }
+    arm_connection_event_listener((NabtoClientConnection*)data);
+}
+
+void init_connection_event_listener(NabtoClient* context, NabtoClientConnection* connection) {
+    listener_ = nabto_client_listener_new(context);
+    eventFuture_ = nabto_client_future_new(context);
+    connectionEvent_ = -1;
+    nabto_client_connection_events_init_listener(connection, listener_);
+}
+
+void arm_connection_event_listener(NabtoClientConnection* connection) {
+    // initialize future for the event listener, the future is resolved when a new connection event is
+    // ready or the listener has been stopped (also see https://docs.nabto.com/developer/platforms/embedded/nabto_futures.html#listeners)
+    nabto_client_listener_connection_event(listener_, eventFuture_, &connectionEvent_);
+    nabto_client_future_set_callback(eventFuture_, connection_event_cb, connection);
+}
+
+// connection event specifics
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+
+void connected_cb(NabtoClientFuture* future, NabtoClientError ec, void* data) {
+    NabtoClientConnection* connection = (NabtoClientConnection*)data;
+    char* f;
+    if (ec != NABTO_CLIENT_EC_OK) {
+        print_connect_error(ec, connection);
+    } else if (nabto_client_connection_get_device_fingerprint_hex(connection, &f) != NABTO_CLIENT_EC_OK) {
```

Comment:
The example is about handling connection events, but then we choose to handle the connected case without using a connection event. It would be more clean for the demo purpose to handle the error case here, but the success case in connection_event_cb

Reply:
I disagree, let's discuss in a meeting. The intuitive flow imho is "open connection, use connection" - similar to "open stream, use stream" and "open tunnel, use tunnel" etc. Moving the "use connection" to connection event handling breaks this pattern - connection events are currently a "FYI" feature specific for connections. Then we should introduce something similar for streams, tunnels and coap.